deployments/prod/environment.py

```diff
@@ -5,6 +5,319 @@
 )
 
 
+def partition_prefix_length(n: int) -> int:
+    """
+    For a given number of subgraphs, return a partition prefix length that
+    yields at most 512 subgraphs per partition.
+
+    >>> [partition_prefix_length(n) for n in (0, 1, 512, 513, 16 * 512, 16 * 513 )]
+    [0, 0, 0, 1, 1, 2]
+    """
+    return 1 + partition_prefix_length(n // 16) if n > 512 else 0
+
+
+ma = 1  # managed access
+pop = 2  # remove snapshot
+
+
+def mksrc(google_project, snapshot, subgraphs, flags: int = 0):
+    # FIXME: Use regular expression to split snapshot name
+    #        https://github.com/DataBiosphere/azul/issues/3912
+    try:
+        _, env, project, _, ds_date, ds_qual, ss_date, ss_qual = snapshot.split('_')
+    except ValueError:
```

Comment:
todo: split on the components that are always present, we only really need the project.